From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Sun, 2 Aug 2026 13:37:05 +0200
Subject: media: hm5065: Report control changes as events

A driver whose firmware moves control values on its own - the AF
status, the exposure the AEC converged on - is exactly the case the
control event mechanism exists for. Subscribers now hear about
control changes instead of having to poll, and log_status dumps the
handler for debugging, same as the gc2145.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/media/i2c/hm5065.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/hm5065.c b/drivers/media/i2c/hm5065.c
index 111111111111..222222222222 100644
--- a/drivers/media/i2c/hm5065.c
+++ b/drivers/media/i2c/hm5065.c
@@ -29,6 +29,7 @@
 #include <media/v4l2-async.h>
 #include <media/v4l2-cci.h>
 #include <media/v4l2-ctrls.h>
+#include <media/v4l2-event.h>
 #include <media/v4l2-device.h>
 #include <media/v4l2-fwnode.h>
 #include <media/v4l2-subdev.h>
@@ -1840,6 +1841,9 @@ static int hm5065_s_register(struct v4l2_subdev *sd,
 /* }}} */
 
 static const struct v4l2_subdev_core_ops hm5065_core_ops = {
+	.log_status = v4l2_ctrl_subdev_log_status,
+	.subscribe_event = v4l2_ctrl_subdev_subscribe_event,
+	.unsubscribe_event = v4l2_event_subdev_unsubscribe,
 #ifdef CONFIG_VIDEO_ADV_DEBUG
 	.g_register = hm5065_g_register,
 	.s_register = hm5065_s_register,
@@ -1969,7 +1973,8 @@ static int hm5065_probe(struct i2c_client *client)
 	v4l2_i2c_subdev_init(&sensor->sd, client, &hm5065_subdev_ops);
 	sensor->sd.internal_ops = &hm5065_internal_ops;
 
-	sensor->sd.flags = V4L2_SUBDEV_FL_HAS_DEVNODE;
+	sensor->sd.flags = V4L2_SUBDEV_FL_HAS_DEVNODE |
+			   V4L2_SUBDEV_FL_HAS_EVENTS;
 	sensor->pad.flags = MEDIA_PAD_FL_SOURCE;
 	sensor->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
 	ret = media_entity_pads_init(&sensor->sd.entity, 1, &sensor->pad);
-- 
Armbian

